### PR TITLE
SONARJAVA-2194 remove BinaryRelation

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

**Comment by saberduck** [comment]
This seems big, but basically it's just copying code from BinaryRelation to RelationalSymbolicValue with corresponding change of type. Tests needed to be modified to be able to produce "not normalized" String representation of relation.

File: java-frontend/src/main/java/org/sonar/java/se/SymbolicExecutionVisitor.java [hunk 1]
```diff
@@ -76,7 +76,7 @@ public MethodBehavior execute(MethodTree methodTree) {
       } else {
         return walker.visitMethod(methodTree);
       }
-    } catch (ExplodedGraphWalker.MaximumStepsReachedException | ExplodedGraphWalker.ExplodedGraphTooBigException | BinaryRelation.TransitiveRelationExceededException exception) {
+    } catch (ExplodedGraphWalker.MaximumStepsReachedException | ExplodedGraphWalker.ExplodedGraphTooBigException | RelationalSymbolicValue.TransitiveRelationExceededException exception) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/SymbolicExecutionVisitor.java]: MINOR Split this 187 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 2]
```diff
@@ -209,42 +198,202 @@ RelationalSymbolicValue inverse() {
...
+      return null;
+    }
+
+    return new RelationalSymbolicValue(other.kind,
+      hasOperand(other.leftOp) ? differentOperand(other) : other.leftOp,
+      hasOperand(other.leftOp) ? other.rightOp : differentOperand(other));
+  }
+
+  @CheckForNull
+  private RelationalSymbolicValue lessThanTransitiveBuilder(RelationalSymbolicValue other) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: MAJOR Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1142)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/SymbolicExecutionVisitor.java [hunk 3]
```diff
@@ -76,7 +76,7 @@ public MethodBehavior execute(MethodTree methodTree) {
       } else {
         return walker.visitMethod(methodTree);
       }
-    } catch (ExplodedGraphWalker.MaximumStepsReachedException | ExplodedGraphWalker.ExplodedGraphTooBigException | BinaryRelation.TransitiveRelationExceededException exception) {
+    } catch (ExplodedGraphWalker.MaximumStepsReachedException | ExplodedGraphWalker.ExplodedGraphTooBigException | TransitiveRelationExceededException exception) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/SymbolicExecutionVisitor.java]: Why not `RelationalSymbolicValue.TransitiveRelationExceededException` ?

File: java-frontend/src/main/java/org/sonar/java/se/ProgramState.java [hunk 4]
```diff
@@ -24,6 +24,7 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Lists;
+
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramState.java]: useless change! :dagger:

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 5]
```diff
@@ -209,42 +198,202 @@ RelationalSymbolicValue inverse() {
...
-      .map(r -> r.deduceTransitiveOrSimplified(relation))
-      .filter(Objects::nonNull)
-      .map(RelationalSymbolicValue::binaryRelationToSymbolicValue)
-      .collect(Collectors.toList());
+  RelationState resolveState(ProgramState programState) {
+    if (hasSameOperand()) {
+      return relationStateForSameOperand();
+    }
+
+    Set<RelationalSymbolicValue> allRelations = knownRelations(programState).collect(Collectors.toCollection(HashSet::new));
```
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: No, order is not relevant here, we are testing all relations with current relation, if any of them resolves `this` we return the result

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java [hunk 6]
```diff
@@ -209,42 +198,202 @@ RelationalSymbolicValue inverse() {
...
+    return knownRelations(programState)
+      .map(this::deduceTransitiveOrSimplified)
+      .filter(Objects::nonNull)
+      .collect(Collectors.toList());
+  }
+
+  private static Stream<RelationalSymbolicValue> knownRelations(ProgramState programState) {
+    return programState.getValuesWithConstraints(BooleanConstraint.TRUE)
+      .stream()
+      .filter(sv -> sv instanceof RelationalSymbolicValue)
```
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: I am not really convinced that it is more readable, but OK

File: java-frontend/src/main/java/org/sonar/java/se/ProgramState.java (hunk 4 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramState.java]: sorry, didn't notice

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java (hunk 5 above)
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java]: Okay.

File: java-frontend/src/main/java/org/sonar/java/se/ProgramState.java (hunk 4 above)
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/ProgramState.java]: And so not changed?

File: java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationState.java [hunk 7]
```diff
@@ -23,9 +23,9 @@
 
 /**
  * This enum values are the possible returns of the method implies(SymbolicValueRelation) of class SymbolicValueRelation.
- * @see BinaryRelation#implies
+ * @see RelationalSymbolicValue#implies(RelationalSymbolicValue)
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationState.java]: There is still a reference in the javadoc of [RelationStateTable.javaL#76](https://github.com/SonarSource/sonar-java/blob/master/java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationStateTable.java#L76). Please update it.

**Pull request merged**